### ISSUE TITLE
tarfile-unsafe-members (S202) is emitted for ZipFile.extractall

### ISSUE THREAD

**Issue by lucidfrontier45** [astral-sh/ruff#22821]
### Summary

According to https://docs.astral.sh/ruff/rules/tarfile-unsafe-members/ this rule should only be applied to tarfile but actually it is also emitted for zipfile. It only happens if I also imported `tarfile` module.

NG
```py
import tarfile
import zipfile
from pathlib import Path

asset_path = Path("path/to/your/archive.zip")
extract_dir = Path("path/to/extract/directory")


with zipfile.ZipFile(asset_path, "r") as zf:
    zf.extractall(extract_dir) <- S202 warning
```

OK
```py
import zipfile
from pathlib import Path

asset_path = Path("path/to/your/archive.zip")
extract_dir = Path("path/to/extract/directory")


with zipfile.ZipFile(asset_path, "r") as zf:
    zf.extractall(extract_dir) <- no warning
```

### Version

0.14.14

**Comment by ntBre** [comment]
Thanks for the report!

Your observation is basically identical to how the rule works. It checks if `tarfile` is imported:

https://github.com/astral-sh/ruff/blob/da8ee78491f784cdead4c69e83c58bc87fd0622c/crates/ruff_linter/src/rules/flake8_bandit/rules/tarfile_unsafe_members.rs#L53

and if the attribute is `extractall`:

https://github.com/astral-sh/ruff/blob/da8ee78491f784cdead4c69e83c58bc87fd0622c/crates/ruff_linter/src/rules/flake8_bandit/rules/tarfile_unsafe_members.rs#L60

I think a similar pattern is used in a lot of the `S` rules, but we could try to improve this by resolving the attribute value to an assignment or using our limited type inference abilities.

https://play.ruff.rs/69afb8e3-90ae-4d6e-afe1-24bc513b7ce7

**Comment by ntBre** [comment]
After thinking about #22880 for a while, I think these rules intentionally bias toward false positives. Ideally we would infer the type of the `extractall` receiver and confirm that it's a `TarFile`, but I think our existing type inference capabilities are too limited to do that reliably for these security-focused rules. I'll add the https://github.com/astral-sh/ruff/labels/type-inference label to revisit this when we have access to more of ty's abilities.
